src/sparse/linalg/ordering.rs

```diff
@@ -13,74 +13,223 @@ pub struct Ordering<I> {
+    }
+}
+
+/// Computes the rooted level structure rooted at `root`,
+/// returning the index of vertex of the last level with minimum degree, called "contender", and the height of the rls.
+fn rls_contender_and_height<N, I, Iptr>(
+    root: usize,
+    degrees: &[usize],
+    mat: &CsMatViewI<N, I, Iptr>,
+) -> (usize, usize)
+where
+    N: PartialEq,
+    I: SpIndex,
+    Iptr: SpIndex,
+{
+    // One might wonder: "Why are we not reusing the rooted level structure (rls) we build here,
+    // isn't it basically the same thing we build in the rcm again, afterwards?""
+    // The answer: Yes, but, No.
+    //
+    // The rooted level structure here differs from the one built afterwards by its order.
+    // The required order is very nasty indeed: the position of a vertex in its level depends primarily on the position of
+    // its neighboring vertex in the previous level, and secondarily on its degree.
+    //
+    // Still, one may think: "Well, then just keep the rls around, and sort it if its root is choosen as starting vertex".
+    // This is easier said than done, let's consider some strategies to do that:
+    // 1. Sort it without any additionally stored information. That would require going through the entire vec again, one by one.
+    //    This would erase the overhead of allocating and then deallocating the rls, but making this pass performant
+    //    requires non-trivial, error-prone code. Overall, this strategies perfomance gains can at most be minimal.
+    // 2. Store additional information, like the delimeters of levels, neighbouring vertex delimeters, etc.
+    //    That would require doing additional work (computing and storing the information) while building the rls,
+    //    and does not speed up sorting afterwards significantly, as the levels still need to be sorted serially.
+    //    Overall, this strategy comes at a significant cost in memory, and it's performance improvements are debatable at best.
+    // 3. Maybe just build any rls in a way that makes it a valid rcm odering?
+    //    That would be optimal if we always find a pseudoperipheral vertex on first try. Unfortunately, we rarely do,
+    //    typical are a few swaps, meaning this strategy, overall, comes with a loss of performance.
+    //
+    // So, thats why we discard the rls. One may feel free to try on his own.
+
+    let nb_vertices = degrees.len();
+
+    // This is ok, if we are given a valid root we can never reach an invalid vertex.
+    let mut visited = vec![false; nb_vertices];
+
+    let mut rls = Vec::with_capacity(nb_vertices);
+
+    // Start out by pushing the root.
+    visited[root] = true;
+    rls.push(root);
+
+    let mut rls_index = 0;
+
+    // For calculating the height.
+    let mut height = 0;
+    let mut current_level_countdown = 1;
+    let mut next_level_countup = 0;
+
+    // The last levels len is used to compute the contender in the end.
+    let mut last_level_len = 1;
+
+    while rls_index < rls.len() {
```

Comment:
Isn't this actually a for loop `for rls_index in 0..rls.len()` ? Or maybe I'm missing something.

Reply:
My bad, I missed that `rls` was pushed to during the loop.